include/swift/IRGen/SwiftIRDetailsCollector.h

```diff
@@ -0,0 +1,61 @@
+//===--- SwiftIRDetailsCollector.h - Get ABI details for decls --*- C++ -*-===//
```

Comment:
I wonder if we could just expose the type info from IRGen? It would be a big-ish refactor, but maybe we could just factor that API out of IRGen? Thoughts?

Reply:
I was thinking about that, but I would need to pull out a lot of logic out of IRGen right now, and it's unclear how much I actually need just to get the layout and ABI right. Perhaps it's the right thing to do, but it's not obvious to me just yet. This is something I can discuss with @rjmccall in our sync up. For now I can actually refactor this to pull out specific query interface instead of having a visitor .

Reply:
We decided to provide a high level interface for now that's pretty much decoupled from IRGen / SILGen but implemented using IRGen by setting up the required context, so that we can avoid depending on running SILGen/IRGen in order to emit the header.

Reply:
I will update the patch to reflect that.